whats-that-typeclass-alternative/article.md

```diff
@@ -0,0 +1,398 @@
+You may come across a programming task where you want to implement something like:
+
+Try _this_ and if it doesn't work out try _that_ - or one could say a computation composed of two (or more) alternatives which should be tried
+one after after the other until one of them produces as useful result.
+
+As a first example lets build a function which takes a `String` as input and returns `Just String` if that input is either `"bob"` or `"alice"`.
+
+Although there might be a lot of other ways to solve this task, we do so defining 2 functions:
+
+```haskell
+isAlice :: String -> Maybe String
+isAlice a = if a == "alice" then Just a else Nothing
+
+isBob :: String -> Maybe String
+isBob a = if a == "bob" then Just a else Nothing
+```
+
+We could then composed those two function into a single one f.e. like this:
+
+```haskell
+isAliceOrBob :: String -> Maybe String
+isAliceOrBob a =
+  case isAlice a of
+    Just _ -> Just a
+    Nothing -> case isBob a of
+                 Just _ -> Just a
+                 Nothing -> Nothing
+```
+
+```none
+λ> isAliceOrBob "alice"
+Just "alice"
+λ> isAliceOrBob "bob"
+Just "bob"
+λ> isAliceOrBob "charly"
+Nothing
+```
+
+This works well but will become ugly soon when we want to add more alternatives to the computation  f.e. a `isCharliy :: String -> Maybe String`.
+
+It would be nice if we would have a operator to write this in a nice way like:
+
+```haskell
+isCharly :: String -> Maybe String
+isCharly a = if a == "charly" then Just a else Nothing
+
+isAliceOrBobOrCharly :: String -> Maybe String
+isAliceOrBobOrCharly a = isAlice a <|> isBob a <|> isCharly a
+-- or
+isAliceOrBobOrCharly a = isAliceOrBob a <|> isCharly a
+```
+
+
+## The Alternative typeclass
+
+Actually the above code is valid haskell an all you have to do to make use of the function `<|>` is to import it from `Control.Aplicative`:
+
+```none
+λ> :info (<|>)
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  ...
+  (<|>) :: f a -> f a -> f a
+  ...
+  	-- Defined in ‘GHC.Base’
+infixl 3 <|>
+```
+
+It is defined in terms of the `Alternative` typeclass. So get some more information about that typeclass:
+
+```haskell
+type Alternative :: (* -> *) -> Constraint
+class Applicative f => Alternative f where
+  empty :: f a
+  (<|>) :: f a -> f a -> f a
+  some :: f a -> f [a]
+  many :: f a -> f [a]
+  {-# MINIMAL empty, (<|>) #-}
+  	-- Defined in ‘GHC.Base’
+instance Alternative ZipList -- Defined in ‘Control.Applicative’
+instance Alternative [] -- Defined in ‘GHC.Base’
+instance Alternative Maybe -- Defined in ‘GHC.Base’
+instance Alternative IO -- Defined in ‘GHC.Base’
+```
+
+This already gives us alot of useful information about that typeclass:
+
+1. A type implementing `Alternative` must be of kind `* -> *`
+2. It must also implement `Applicative` (and with that also `Functor`)
+3. To implement it you must implement `<|>` and `empty`
+4. Next to those 2 functions it gives you `some` and `many` as well
```

Comment:
You don't really cover what these do, I think it would be beneficial.

Reply:
added a note that I will explain that further below